bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java

```diff
@@ -552,27 +554,37 @@ private Object retrievePrimitiveValue(TomlNode tomlValue, String variableName, T
         return getBalValue(variableName, type, value);
     }
 
-    private BString retrieveUnionValue(TomlNode tomlValue, String variableName, UnionType unionType) {
-        TomlNode tomlNode = ((TomlKeyValueNode) tomlValue).value();
-        visitedNodes.add(tomlValue);
-        if (tomlNode.kind() != getEffectiveTomlType(unionType, variableName)) {
-            invalidTomlLines.add(tomlValue.location().lineRange());
-            throw new ConfigException(CONFIG_INCOMPATIBLE_TYPE, getLineRange(tomlValue), variableName, unionType,
+    private Object retrieveUnionValue(TomlNode tomlValue, String variableName, BUnionType unionType) {
+        Object balValue = Utils.getBalValue(tomlValue, visitedNodes);
+        List<Type> convertibleTypes = new ArrayList<>();
+        for (Type type : unionType.getMemberTypes()) {
+            if (TypeChecker.checkIsLikeType(balValue, type, false)) {
+                convertibleTypes.add(type);
+            }
+        }
+        if (convertibleTypes.isEmpty()) {
+            throw new ConfigException(CONFIG_INCOMPATIBLE_TYPE, getLineRange(tomlValue), variableName,
```

Comment:
Shall we add this before throwing the throwing ConfigException? (in other places as well)
```ballerina
   invalidTomlLines.add(tomlValue.location().lineRange());
```
This is to prevent from duplicated `unused configuration value` warnings.